README.md
# 💡A-Chika
![npm version](https://img.shields.io/npm/v/achika?style=flat-square)
[![Dependencies](https://img.shields.io/david/nandenjin/achika?style=flat-square)](https://david-dm.org/nandenjin/achika)
[![devDependencies](https://img.shields.io/david/dev/nandenjin/achika?style=flat-square)](https://david-dm.org/nandenjin/achika)
[![Maintainability](https://api.codeclimate.com/v1/badges/8f649adb46919e5838fa/maintainability?style=flat-square)](https://codeclimate.com/github/nandenjin/achika/maintainability)
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](LICENSE)


Test tool for Art-Net receivers.

## Features
- Transmit Art-Net blink signals from command line
- Options
  - Universe
  - Multiple channels
  - Blink interval

## Install & Quick use
```
$ npm install -g achika
$ achika 1

// OR just

$ npx achika 1
```

## Usage
```
Usage:
  $ achika -h [HOST IP] [ADDRESS1] [ADDRESS2] [...]

  --version, -v Show version
  --host, -h [HOST IP] Set host IP
  --universe, -u [UNIVERSE] Set universe
  --interval, -i  [INTERVAL SEC] Set interval in seconds
  --max [MAX VALUE] / --min [MIN VALUE] Set max/min value
```

## Contributing
Issues and PRs for adding features, bug fix or refactoring are welcome. Please do not include built files in your PRs.

## Naming
Named from `L-Chika（Lチカ）`, the Japanese word of simple "Hello World" in electric circuits development, just blink a LED.

In this case `L` comes from "LED", so `A` of "A-Chika" comes from "Art-Net".


CHANGELOG.md
## [1.1.5](https://github.com/nandenjin/achika/compare/v1.1.4...v1.1.5) (2021-05-09)



## [1.1.4](https://github.com/nandenjin/achika/compare/v1.1.3...v1.1.4) (2020-12-14)



## [1.1.3](https://github.com/nandenjin/achika/compare/v1.1.2...v1.1.3) (2020-11-01)



## [1.1.2](https://github.com/nandenjin/achika/compare/v1.1.1...v1.1.2) (2020-10-02)





src/chika.js
const chalk = require('chalk')
const consola = require('consola')
const ArtNet = require('artnet')

function chika(
  { host, universe = 1, interval, min, max, version, sendAll, refreshInterval },
  { input, showHelp, showVersion }
) {
  if (version) showVersion()

  universe /= 1
  interval /= 1

  refreshInterval /= 1

  if (sendAll) {
    consola.info('sendAll flag specified. Transmitting all channels...')
  }

  const artnet = ArtNet({
    host,
    universe,
    refresh: refreshInterval,
  })

  const addrs = []

  if (input.length === 0) {
    consola.error('No address specified')
    showHelp()
  }

  input.forEach(ins => {
    ins.split(',').forEach(a => {
      if (a.match(/^([0-9]+)-([0-9]+)$/)) {
        const startAddr = Math.min(+RegExp.$1, +RegExp.$2)
        const endAddr = Math.max(+RegExp.$1, +RegExp.$2)

        for (let i = startAddr; i <= endAddr; i++) {
          addrs.push(i)
        }
      } else if (a.match(/^[0-9]+$/)) {
        addrs.push(a / 1)
      } else {
        consola.error('Invalid address')
        showHelp()
      }
    })
  })

  consola.log(`Universe: ${universe}`)
  consola.log(`Host: ${host}`)

  let vs = new Array(addrs.length).fill(false)
  let csr = 0

  setInterval(() => {
    vs[csr] = !vs[csr]
    csr++
    if (csr >= vs.length) csr = 0

    let output =
      '\r' +
      addrs
        .map((a, i) => {
          const addrStr = ('   ' + a).slice(-3)
          if (vs[i]) return chalk.bgWhite(chalk.black(addrStr))
          else return chalk.black(chalk.white(addrStr))
        })
        .join(' ')

    process.stdout.write(output)

    vs.forEach((v, i) => artnet.set(addrs[i], v ? max : min))
  }, interval * 1000)
}

module.exports = chika


src/chika.ts

import chalk from 'chalk'
import consola from 'consola'
import ArtNet from 'artnet'

export function chika({
  host,
  universe = 1,
  interval,
  min,
  max,
  version,
  sendAll,
  refreshInterval
}, {
  input,
  showHelp,
  showVersion
}) {

  if(version) showVersion()

  universe /= 1
  interval /= 1

  refreshInterval /= 1

  if(sendAll) {
    consola.info('sendAll flag specified. Transmitting all channels...')
  }

  const artnet = ArtNet({
    host,
    universe,
    refresh: refreshInterval
  })

  const addrs = [];

  if(input.length === 0) {
    consola.error('No address specified')
    showHelp()
  }

  input.forEach(ins => {
    ins.split(',')
      .forEach(a => {

        if(a.match(/^([0-9]+)\-([0-9]+)$/)) {

          const startAddr: number = Math.min(+RegExp.$1, +RegExp.$2)
          const endAddr: number = Math.max(+RegExp.$1, +RegExp.$2)

          for(let i = startAddr; i <= endAddr; i++) {
            addrs.push(i)
          }

        }else if(a.match(/^[0-9]+$/)) {

          addrs.push(a/1)

        }else {

          consola.error('Invalid address')
          showHelp()

        }
      })
  })

  consola.log(`Universe: ${universe}`)
  consola.log(`Host: ${host}`)

  let vs: boolean[] = new Array(addrs.length).fill(false)
  let csr: number = 0

  setInterval(() => {

    vs[csr] = !vs[csr]
    csr++
    if(csr >= vs.length) csr = 0;

    let output = '\r' + addrs.map((a, i)=>{

      const addrStr = ('   ' + a).slice(-3)
      if(vs[i]) return chalk.bgWhite(chalk.black(addrStr))
      else return chalk.black(chalk.white(addrStr))

    }).join(' ')

    process.stdout.write(output)

    vs.forEach((v, i) => artnet.set(addrs[i], v ? max : min))

  }, interval * 1000)

}



src/index.ts

import meow from 'meow'
import {chika} from './chika'

const cli = meow(`
Usage:
  $ achika -h [HOST IP] [ADDRESS1] [ADDRESS2] [...]

  --version, -v Show version
  --host, -r [HOST IP] Set host IP
  --universe, -u [UNIVERSE] Set universe
  --interval, -i  [INTERVAL SEC] Set interval in seconds
  --max [MAX VALUE] / --min [MIN VALUE] Set max/min value
`, {
  flags: {

    version: {
      type: 'boolean',
      alias: 'v'
    },

    host: {
      type: 'string',
      alias: 'h',
      default: '255.255.255.255'
    },

    universe: {
      type: 'string',
      alias: 'u',
      default: 1
    },

    interval: {
      type: 'string',
      alias: 'i',
      default: 1
    },

    max: {
      type: 'string',
      default: 255
    },

    min: {
      type: 'string',
      default: 0
    },

    sendAll: {
      type: 'boolean',
      default: false
    },

    refreshInterval: {
      type: 'string',
      default: 1000/60
    }

  }
} );

chika(cli.flags, cli);


src/index.js
#! /usr/bin/env node

const meow = require('meow')
const chika = require('./chika')

const cli = meow(
  `
Usage:
  $ achika -h [HOST IP] [ADDRESS1] [ADDRESS2] [...]

  --version, -v Show version
  --host, -h [HOST IP] Set host IP
  --universe, -u [UNIVERSE] Set universe
  --interval, -i  [INTERVAL SEC] Set interval in seconds
  --max [MAX VALUE] / --min [MIN VALUE] Set max/min value
`,
  {
    flags: {
      version: {
        type: 'boolean',
        alias: 'v',
      },

      host: {
        type: 'string',
        alias: 'h',
        default: '255.255.255.255',
      },

      universe: {
        type: 'string',
        alias: 'u',
        default: '1',
      },

      interval: {
        type: 'string',
        alias: 'i',
        default: '1',
      },

      max: {
        type: 'string',
        default: '255',
      },

      min: {
        type: 'string',
        default: '0',
      },

      sendAll: {
        type: 'boolean',
        default: false,
      },

      refreshInterval: {
        type: 'string',
        default: 1000 / 60 + '',
      },
    },
  }
)

chika(cli.flags, cli)
